src/unit.c

```diff
@@ -57,114 +50,21 @@ SEXP unit(double value, int unit)
  * (i.e., NOT unitLists or unitArithmetics)
  */
 double unitValue(SEXP unit, int index) {
-    /* Recycle values if necessary (used in unit arithmetic)
-     */
-    int n = LENGTH(unit);
-    return numeric(unit, index % n);
+  return Rf_asReal(VECTOR_ELT(VECTOR_ELT(unit, index % unitLength(unit)), 0));
 }
 
 int unitUnit(SEXP unit, int index) {
-    SEXP units = getAttrib(unit, install("valid.unit"));
-    /* Recycle units if necessary 
-     */
-    int n = LENGTH(units);
-    return INTEGER(units)[index % n];
+  return Rf_asInteger(VECTOR_ELT(VECTOR_ELT(unit, index % unitLength(unit)), 2));
 }
 
 SEXP unitData(SEXP unit, int index) {
-    SEXP result;
-    SEXP data = getAttrib(unit, install("data"));
-    if (isNull(data))
-	result = R_NilValue;
-    else if(TYPEOF(data) == VECSXP) {
-	/* Recycle data if necessary 
-	 */
-	int n = LENGTH(data);
-	result = VECTOR_ELT(data, index % n);
-    } else {
-	warning("unit attribute 'data' is of incorrect type");
-	return R_NilValue;
-    }
-    return result;
+  return VECTOR_ELT(VECTOR_ELT(unit, index % unitLength(unit)), 1);
 }
 
-/* Accessor functions for unit arithmetic object
+/* Old alternative to LENGTH when using that didn't work on all unit struct
```

Comment:
Why keep this function at all?

Reply:
It is called throughout the codebase (and maybe in lattice as well)... seemed easier and less intrusive to just keep it

Reply:
is this bad style, btw?

Reply:
I think you should file an issue to replace `unitLength()` with `LENGTH()` everywhere. One of the things that I think makes base R code hard to read is the simultaneous use of deprecated and new interfaces.

Reply:
It is a problem of timing (I think). It is required for the old spec because of the trinity of implementations, and I can't remove it before the new unit spec has been accepted because it is used outside grid as well